### PR TITLE
LLVM14

### PR DESCRIPTION
**Pull request by apmasell**

### CODE REVIEW

**Comment by sklam** [comment]
https://github.com/numba/numba/issues/8738 documents why we cannot use llvm14 on linux-aarch64

File: buildscripts/condarecipe.local/meta.yaml [hunk 1]
```diff
@@ -34,7 +34,7 @@ requirements:
     - setuptools
     - importlib_metadata       # [py<39]
     # On channel https://anaconda.org/numba/
-    - llvmlite >=0.40.0dev0,<0.40
+    - llvmlite >=0.40.0dev0llvm14,<0.40
```
Comment (sklam) [inline buildscripts/condarecipe.local/meta.yaml:37]: Do we need to split build of llvmlite on aarch64 to use llvm11? Or, can we just say we stay on llvmlite 0.39? This is something we should decide @numba/fpoc

**Review by sklam — COMMENTED** [review]
Just one suggestion on adding a test for fanout_raise refpruning. Beside needing decision on how to move forward for llvm14/11 split builds. Everything else looks good.

**Comment by sklam** [comment]
I tested this patch with LLVM11 (with _ensure_llvm disabled) on x86 as a guess-estimate of problems on linux-aarch64. Ignoring failures that are testing llvmlite version and tests that are specific to x64. Everything else is ok.

File: buildscripts/condarecipe.local/meta.yaml (hunk 1 above)
Comment (stuartarchibald) [reply buildscripts/condarecipe.local/meta.yaml:37]: xref #8756

File: numba/tests/test_debuginfo.py [hunk 2]
```diff
@@ -223,9 +223,16 @@ def foo(a):
...
-        self.assertIn(''.join(op_expect), ''.join(op_seq))
+        op_expect = {'fadd', 'fmul', 'fdiv'}
+        started = False
+        for x in instrs:
+            if x.opcode in op_expect:
+                op_expect.remove(x.opcode)
+                if not started:
+                    started = True
+            elif op_expect and started:
+                self.assertTrue(False, "Math opcodes are not contiguous")
```
Comment (stuartarchibald) [inline numba/tests/test_debuginfo.py:234]:
```suggestion
                self.fail("Math opcodes are not contiguous")
```

File: numba/__init__.py [hunk 3]
```diff
@@ -143,7 +143,7 @@ def test(argv, **kwds):
 
 
 _min_llvmlite_version = (0, 40, 0)
-_min_llvm_version = (11, 0, 0)
+_min_llvm_version = (14, 0, 0)
```
[the remark this answers is not in the document]
Comment (apmasell) [reply numba/__init__.py:146]: I think should be just 11. We're theoretically fine with LLVM 11 on any platform, right?

[the remark this answers is not in the document]
Comment (sklam) [reply numba/__init__.py:146]: sure

**Comment by sklam** [comment]
@apmasell, this PR needs a merge main to resolve conflicts

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/core/codegen.py [hunk 4]
```diff
@@ -1215,14 +1215,19 @@ def _module_pass_manager(self, **kwargs):
...
+                # this interacted badly with the existing use of the
+                # InstructionCombiner here and ended up with PHI nodes that
+                # prevented vectorization from working. The desired
+                # vectorization effects can be achieved with this in LLVM 11
+                # (and also < 11) but at a potentially slightly higher cost:
+                pm.add_licm_pass()
+                pm.add_cfg_simplification_pass()
+            else:
+                pm.add_instruction_combining_pass()
+                pm.add_jump_threading_pass()
```
Comment (stuartarchibald) [inline numba/core/codegen.py:1229]: Please could a comment be added explaining what the problem was and why this fixes it? Thanks!

**Comment by sklam** [comment]
I tested this branch against the latest llvmlite package at https://github.com/numba/numba/pull/8807 and it passed (except windows are failing due to timeout)

File: numba/tests/test_refop_pruning.py [hunk 5]
```diff
@@ -118,6 +118,22 @@ def func(n):
                     raise ValueError
             return x
 
+        with set_refprune_flags('per_bb,fanout'):
+            self.check(func, (types.intp), basicblock=True, diamond=False,
+                       fanout=True, fanout_raise=False)
+
+    def test_fanout_3(self):
```
Comment (sklam) [inline numba/tests/test_refop_pruning.py:125]: Do we need to skip this for linux-aarch64 or just when llvm version == 11?

**Review by sklam — COMMENTED** [review]
A few more comments need address before merging, including https://github.com/numba/numba/pull/8535/files#r1100560473
and mainly reverting the change in the llvmlite version.

**Review by sklam — APPROVED** [review]

**Comment by sklam** [comment]
Builds are failing expectedly because they are picking up the old llvmlite build

**Pull request merged**